include/swift/AST/DiagnosticConsumer.h

```diff
@@ -136,32 +136,110 @@ class NullDiagnosticConsumer : public DiagnosticConsumer {
-  using ConsumerPair =
-      std::pair<std::string, std::unique_ptr<DiagnosticConsumer>>;
+  /// be associated with.
+  class Subconsumer {
+    friend std::unique_ptr<DiagnosticConsumer>
+    FileSpecificDiagnosticConsumer::consolidateSubconsumers(
+        SmallVectorImpl<Subconsumer> &subconsumers);
+
+    /// The name of the input file that a consumer and diagnostics should
+    /// be associated with. An empty string means that a consumer is not
+    /// associated with any particular buffer, and should only receive
+    /// diagnostics that are not in any of the other consumers' files.
+    std::string inputFileName;
+
+    /// The consumer (if any) for diagnostics associated with the inputFileName.
+    /// A null pointer for the DiagnosticConsumer means that diagnostics for
+    /// this file should not be emitted.
+    std::unique_ptr<DiagnosticConsumer> consumer;
+
+    // Has this subconsumer ever handled a diagnostic that is an error?
+    bool hasAnErrorBeenConsumed = false;
+
+  public:
+    std::string getInputFileName() const { return inputFileName; }
+
+    DiagnosticConsumer *getConsumer() const { return consumer.get(); }
+
+    Subconsumer(std::string inputFileName,
+                std::unique_ptr<DiagnosticConsumer> consumer)
+        : inputFileName(inputFileName), consumer(std::move(consumer)) {}
+
+    void handleDiagnostic(SourceManager &SM, SourceLoc Loc,
+                          DiagnosticKind Kind,
+                          StringRef FormatString,
+                          ArrayRef<DiagnosticArgument> FormatArgs,
+                          const DiagnosticInfo &Info) {
+      if (!getConsumer())
+        return; // Suppress non-primary diagnostic in batch mode.
+      hasAnErrorBeenConsumed |= Kind == DiagnosticKind::Error;
+      getConsumer()->handleDiagnostic(SM, Loc, Kind, FormatString, FormatArgs,
+                                      Info);
+    }
+    
+    void informDriverOfIncompleteBatchModeCompilation() {
+      if (!hasAnErrorBeenConsumed && getConsumer())
+        getConsumer()->informDriverOfIncompleteBatchModeCompilation();
+    }
+  };
 
 private:
   /// All consumers owned by this FileSpecificDiagnosticConsumer.
-  const SmallVector<ConsumerPair, 4> SubConsumers;
+  SmallVector<Subconsumer, 4> Subconsumers;
 
 public:
   // The commented-out consts are there because the data does not change
   // but the swap method gets called on this structure.
-  struct ConsumerSpecificInformation {
+  class ConsumerSpecificInformation {
+  private:
+    /// The range of SourceLoc's for which diagnostics should be directed to
+    /// this subconsumer.
     /*const*/ CharSourceRange range;
-    /// The DiagnosticConsumer may be empty if those diagnostics are not to be
-    /// emitted.
-    DiagnosticConsumer * /*const*/ consumer;
-    bool hasAnErrorBeenEmitted = false;
 
+    /// Index into Subconsumers vector for this subconsumer.
+    /*const*/ unsigned subconsumerIndex;
```

Comment:
Why can't you make these const?

Reply:
Good catch! Back when there was a unique pointer in there, I don't think these things compiled, but I don't recall the details.